### DISCUSSION TITLE
how to maintain focus order on dynamically created form fields

### DISCUSSION THREAD

**Discussion by alicemao23** [react-hook-form/react-hook-form#3310] [Q&A]
So according to documentation on [Controllers](https://react-hook-form.com/api/#Controller]), Input focus orders are based on registration order. I have a form that has a button for the user to add a group of new fields to a subsection of the form. However, on validation, the focus order will focus on the fields that were registered first on load, then go back to focus on the additional field that was dynamically added by user event. Obviously this does not flow visually from the UX perspective. Is there a way for me to manipulate the focus order?

**Comment by bluebill1049** [comment]
yea, not at the moment. I do hear this concern. You will have to convert all input into `Controller` for now. I am planning to introduce a focus order prop which is an array for users to defined the focus order manually to avoid default, but it's still on the paper. You can send a feature request tho, happy to pin on the board and see if many users have similar requests.

**Reply by bluedusk** [reply]
Can we control the focus order now ?

**Reply by Moshyfawn** [reply]
Please, check the whole thread for info before posting your question. The answer to it is in [this comment](https://github.com/react-hook-form/react-hook-form/discussions/3310#discussioncomment-3103375)

**Comment by alicemao23** [comment]
Right, yeah happy to do that! For my usecase though, all of my inputs are Controllers,  it's just that when I add a new set of Controllers based on user action, the new Controllers are prioritized to focus after all the other controllers that were registered on-load. I remember reading somewhere in the documentation that when using useFieldArray, you should pass ref={register()} to make sure that map works correctly to register all the inputs. 

Is there something similar for Controllers?

**Reply by bluebill1049** [reply]
I don't think so. 6.10.0 introduced `ref` in the render props. it should make focus simpler for external controlled component.

**Reply by bluebill1049** [reply]
https://github.com/react-hook-form/react-hook-form/blob/master/CHANGELOG.md

**Comment by alicemao23** [comment]
Okay so just to summarize then. All the inputs in my forms are Controllers, and onFocus for new dynamically added Controller inputs are still being prioritized after the Controllers that were registered when the form first loaded, and there's no way to reorder the registration order?

**Reply by bluebill1049** [reply]
> Controller inputs are still being prioritized after the Controllers that were registered when the form first loaded

can you share a codesandbox with the above statement?

**Reply by alicemao23** [reply]
Yup! https://codesandbox.io/s/react-hook-form-with-ui-library-forked-jp0fe

So in this example above, add a new row after the app has loaded. As you fill out each input, click submit, the form will auto-focus to the next Input that has not passed validations. But it will skip the second row that you've added, goes to the FirstName & LastName inputs, then when those inputs are completed it'll focus to the second row of barcode & price. 

The ideal behavior is that as more rows are added, the onFocus would target each of the barcode price input before advancing to the firstName & lastName fields. 

Thanks for your help btw! 🙏

**Reply by bluebill1049** [reply]
oh, i see it's the focus input within the field array's inputs, we don't have such a feature at the moment.

perhaps you want to include your idea into this feature request as well: https://github.com/react-hook-form/react-hook-form/issues/2155

**Comment by gmwill934** [comment]
Can we manage the order right now? I also have a `Controlled` input and two `Uncontrolled`. Is there a way to manage `focus order` now?

**Reply by Moshyfawn** [reply]
Check out #7523 thread for more info. Also, you can check out [this tweet](https://twitter.com/HookForm/status/1529965338133049345) on a proposed API for RHF API to manage focus order

**Reply by jasonzguo** [reply]
Hello @Moshyfawn 
Is the feature proposed in the tweet being worked on right  now?
